src/corehost/cli/hostpolicy.cpp

```diff
@@ -347,19 +431,55 @@ void trace_hostpolicy_entrypoint_invocation(const pal::string_t& entryPointName)
         entryPointName.c_str());
 }
 
-SHARED_API int corehost_load(host_interface_t* init)
+namespace
 {
-    trace::setup();
+    int corehost_load_internal(host_interface_t* init, bool reinit)
+    {
+        assert(init != nullptr);
+        std::lock_guard<std::mutex> lock{ g_init_lock };
 
-    // Re-initialize global state in case of re-entry
-    g_init = hostpolicy_init_t();
+        // If initialization was already done and reinitialization is requested,
+        // then return immediately.
+        //
+        // [TODO] Instead of checking for reinitialization, verify the existing
+        // supports the currently requested policy.
+        if (g_init_done && !reinit)
+            return StatusCode::Success;
 
-    if (!hostpolicy_init_t::init(init, &g_init))
-    {
-        return StatusCode::LibHostInitFailure;
+        trace::setup();
+
+        // Re-initialize global state in case of re-entry
+        g_init = hostpolicy_init_t{};
+
+        if (!hostpolicy_init_t::init(init, &g_init))
+        {
+            g_init_done = false;
+            return StatusCode::LibHostInitFailure;
+        }
+
+        g_init_done = true;
+        return StatusCode::Success;
     }
-    
-    return 0;
+}
+
+//
+// Loads and initilizes the hostpolicy.
+//
+// The hostpolicy will be reinitialized with the current
+// host_interface_t argument.
+//
+SHARED_API int corehost_load(host_interface_t* init)
+{
+    return corehost_load_internal(init, true /* reinit */);
+}
```

Comment:
I have no idea. This was undefined but the existing code had this logic and I didn't want to change it. I think that @vitek-karas mentioned something about it, but alas none of us know why this was there other than some ominous comment:

```cpp
// Re-initialize global state in case of re-entry
```

Reply:
I have removed this logic since it appears it was written to support a re-entrant test 😖